File: 백준/Silver/15649. N과 M （1）/N과 M （1）.java
import java.util.*;
import java.io.*;




public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int n, m;
    static int[] nums = new int[10];
    static boolean[] used = new boolean[10];

    static void BT(int d) throws IOException {
        if (d == m) {
            for (int i = 0; i < d; ++i){
                bw.write(nums[i] + " ");
            }
            bw.write("\n");
        }

        for (int i = 1; i <= n; ++i) {
            if (used[i]) continue;
            nums[d] = i;
            used[i] = true;
            BT(d + 1);
            used[i] = false;
        }
    }
    public static void main(String[] args) throws IOException {



        int[] input = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        n = input[0];
        m = input[1];

        BT(0);

        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Silver/15649. N과 M （1）/N과 M （1）.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import java.util.*

val br = BufferedReader(InputStreamReader(System.`in`))
val bw = BufferedWriter(OutputStreamWriter(System.`out`))
var n = 0
var m = 0
val buffer = StringBuffer()
lateinit var used: BooleanArray

fun BT(cnt: Int) {
    if (cnt == m) {
        for (ch in buffer) {
            bw.write("$ch ")
        }
        bw.newLine()
    }

    for (i in 1.. n) {
        if (used[i]) continue
        used[i] = true
        buffer.append(i)
        BT(cnt + 1)
        used[i] = false
        buffer.deleteCharAt(buffer.length - 1)
    }
}

fun main() {
    val input = StringTokenizer(br.readLine()," ")
    n = input.nextToken().toInt()
    m = input.nextToken().toInt()
    used = BooleanArray(n + 1) {false}

    BT(0)

    bw.flush()
    bw.close()
    br.close()
}